### DISCUSSION TITLE
synchronized 키워드에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#86] [Java]
미션 진행하면서

- `Random` vs `ThreadLocalRandom`
- `Stack`의 `push()`, `pop()` 

syncronized 키워드 때문에 성능 차이가 있었던 기억이 있어 써봅니다.

**Comment by jimi567** [comment]
synchronized 키워드는 자바에서 멀티스레드 환경에서의 동기화를 위해 사용된다.
멀티스레드 환경에서는 여러 스레드가 동일한 자원을 동시에 접근할 때, 데이터 일관성 문제가 발생할 수 있는데, synchronized 이를 해결할 수 있다.

주요 특징으로는 다음과 같다.
synchronized키워드를 사용하여, 메서드 또는 특정 블록을  크리티컬 섹션(임계 영역)을 지정할 수 있다. 이를 통해 여러 스레드가 동시에 해당 블록이나 메서드를 동시에 접근할 수 없도록 보호할 수 있다.

또한, 내부적으로 모니터 락을 사용하여 동기화를 구현한다. 각 객체는 자신의 모니터 락을 가지고 있으며, 
한 번에 하나의 스레드만 모니터 락을 소유할 수 있도록 한다( = Mutual Exclusion).

멀티스레드 환경에서 데이터 일관성을 유지하는 이점을 가져다 주지만, 과도한 동기화는 성능 저하를 가져올 수 있으므로, 유의해야한다.

**Comment by nayonsoso** [comment]
멀티 스레드를 사용하면 여러 작업을 동시에 처리할 수 있기 때문에 좋은 성능을 낼 수 있지만
스레드간 동기화 문제가 발생할 수 있습니다.
동기화되지 않은 상태에서 여러 쓰레드가 공유 자원에 접근한다면,
안정성을 보장할 수 없어지기 때문입니다.

자바에서는 쓰레드 동기화를 위해 `synchronized` 키워드를 제공합니다.
쓰레드 동기화란, 여러 스레드가 하나의 공유자원에 동시에 접근하지 못하도록 막는 것을 의미합니다.
동기화가 필요한 부분을 임계 영역이라고 하는데요,
synchronized 키워드를 사용해 변수 또는 함수를 임계 영역으로 지정할 수 있습니다.

하지만 synchronized 키워드를 너무 남발하면 성능 저하를 일으킬 수 있습니다.
synchronized를 사용하면 내부적으로 block과 unlock을 처리하게 되는데,
이런 처리들이 곧 성능 저하로 이어질 수 있습니다.

**Reply by seunghye218** [reply]
block 이 blocking 을 뜻하는 걸까요?

**Reply by nayonsoso** [reply]
아뇨, 오타는 아니었고 block을 의도해서 작성한 것입니다.
그런데 찾아보니 상태에 대한 표현은
blocking(locking)과 unlocking이 정확한 표현이라고 합니다.😅

**Comment by GIVEN53** [comment]
멀티 스레드
동기화
blocking
모니터 락
